src/instructlab/model/download.py

```diff
@@ -12,20 +17,250 @@
+        self.download_dest = download_dest
+
+    def _build_oci_model_file_map(self, oci_model_path: str) -> dict:
+        """
+        Helper function to build a mapping between blob files and what they represent
+        """
+        index_hash = ""
+        try:
+            with open(f"{oci_model_path}/index.json", mode="r", encoding="UTF-8") as f:
+                index_ref = json.load(f)
+            match = re.search("sha256:(.*)", index_ref["manifests"][0]["digest"])
+
+            if match:
+                index_hash = match.group(1)
+            else:
+                click.echo(f"could not find hash for index file at: {oci_model_path}")
+                raise click.exceptions.Exit(1)
+        except FileNotFoundError as exc:
+            raise ValueError(f"file not found: {oci_model_path}/index.json") from exc
+        except json.JSONDecodeError as exc:
+            raise ValueError(
+                f"could not read JSON file: {oci_model_path}/index.json"
+            ) from exc
+        except Exception as exc:
+            raise ValueError("unexpected error occurred: {e}") from exc
+
+        try:
+            with open(
+                f"{oci_model_path}/blobs/sha256/{index_hash}",
+                mode="r",
+                encoding="UTF-8",
+            ) as f:
+                index = json.load(f)
+        except FileNotFoundError as exc:
+            raise ValueError(f"file not found: {oci_model_path}/index.json") from exc
+        except json.JSONDecodeError as exc:
+            raise ValueError(
+                f"could not read JSON file: {oci_model_path}/index.json"
+            ) from exc
+        except Exception as exc:
+            raise ValueError("unexpected error occurred: {e}") from exc
+
+        title_ref = "org.opencontainers.image.title"
+        oci_model_file_map = {}
+
+        for layer in index["layers"]:
+            match = re.search("sha256:(.*)", layer["digest"])
+
+            if match:
+                blob_name = match.group(1)
+                oci_model_file_map[blob_name] = layer["annotations"][title_ref]
+
+        return oci_model_file_map
+
+    def download(self):
+        click.echo(
+            f"Downloading model from OCI registry: {self.repository}@{self.release} to {self.download_dest}..."
+        )
+
+        os.makedirs(self.download_dest, exist_ok=True)
+        model_name = self.repository.split("/")[-1]
+        oci_dir = f"{DEFAULTS.OCI_DIR}/{model_name}"
+        os.makedirs(oci_dir, exist_ok=True)
+
+        if which("skopeo") is None:
+            click.secho(
+                "skopeo not detected, but required to perform downloads from OCI registries. Exiting",
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+
+        try:
+            subprocess.run(
+                [
+                    "skopeo",
+                    "copy",
+                    f"{self.repository}:{self.release}",
+                    f"oci:{oci_dir}",
```

Comment:
I think we should use Popen here so that we can:

* log stdout while the download is in progress
* print stderr/stdout when rc is not 0

This might look like:

```python
process = subprocess.Popen(
    [
        "skopeo",
        "copy",
        "docker://quay.io/ai-lab/models/granite-7b-lab:latest",
        f"oci:{oci_dir}",
    ],
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE,
    text=True,
)

# Print stdout line by line as it's being produced
while True:
    output = process.stdout.readline()
    if output == "" and process.poll() is not None:
        break
    if output:
        print(output.strip())

# Check if the process exited with an error
if process.returncode != 0:
    stderr = process.stderr.read()
    print(f"Error: {stderr}", file=sys.stderr)
```

Reply:
possibly because using a PIPE breaks the in-place update that happens to the terminal o/p for the progress bar?

Reply:
Oh that's right, weird, nevermind then, the current code is working better.

We still need to capture stdout/stderr on failure though. Where is it done?

Reply:
from https://docs.python.org/3.5/library/subprocess.html#subprocess.run
```
If check is true, and the process exits with a non-zero exit code, a [CalledProcessError](https://docs.python.org/3.5/library/subprocess.html#subprocess.CalledProcessError) exception will be raised.
```
that's what's being handled right now

Reply:
It also says:

```
The CalledProcessError object will have the return code
    in the returncode attribute, and output & stderr attributes if those streams
    were captured.
```

Today are not captured as far as I can tell.

Reply:
The stdout / stderr are going to the ilab's stdout, stderr which you need for the tty progress bars capability (other than creating a pty would be a bit complex). You can interpret error codes 2 = not found and 1 some other error. A common approach would be "see above previous errors for failure details"